File: pipeline/detectors/make_entering_directory.py
"""
Detector for make entering directory messages.
"""

import re
import typing as T
from pipeline.detectors.base import Detector
from pipeline.models import ErrorClue


class MakeEnteringDirectoryDetector(Detector):
    """
    Detect when make has entered a directory.
    Useful for understanding context for later make errors
    """
    PATTERNS = {
        "make_enter_directory": r"make(?:\[\d+\])?: Entering directory '(?P<directory>[^']+)'"
    }

    EXAMPLES = [
        (
            "make: Entering directory 'helpers'",
            {
                "clue_type": "make_enter_directory",
                "confidence": 1.0,
                "context": {
                    "directory": "helpers",
                },
            },
        ),
    ]

File: src/boil/detectors/make_entering_directory.go
// Detector for make entering directory messages.
package detectors

// MakeEnteringDirectoryDetector detects make entering directory messages
type MakeEnteringDirectoryDetector struct {
	*BaseDetector
}

func NewMakeEnteringDirectoryDetector() (*MakeEnteringDirectoryDetector, error) {
	base, err := NewBaseDetector(
		"MakeEnteringDirectoryDetector",
		100,
		map[string]string{
			"make_enter_directory": `make(?:\[\d+\])?: Entering directory '(?P<directory>[^']+)'`,
		},
		[]DetectorExample{
			{
				Name:     "make_entering_simple",
				Input:    "make: Entering directory '/home/user/project'",
				ClueType: "make_enter_directory",
				Context:  map[string]string{"directory": "/home/user/project"},
			},
			{
				Name:     "make_entering_with_level",
				Input:    "make[1]: Entering directory '/tmp/build'",
				ClueType: "make_enter_directory",
				Context:  map[string]string{"directory": "/tmp/build"},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &MakeEnteringDirectoryDetector{base}, nil
}
